quota/cache/cache_test.go

```diff
@@ -0,0 +1,205 @@
+// limitations under the License.
+
+package cache
+
+import (
+	"context"
+	"errors"
+	"testing"
+	"time"
+
+	"github.com/golang/mock/gomock"
+	"github.com/google/trillian/quota"
+	"github.com/kylelemons/godebug/pretty"
+)
+
+const (
+	batchSize  = 20
+	maxEntries = 10
+)
+
+var (
+	specs = []quota.Spec{
+		{Group: quota.Global, Kind: quota.Read},
+		{Group: quota.Global, Kind: quota.Write},
+	}
+)
+
+func TestCachedManager_GetUser(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	ctx := context.Background()
+	want := "llama"
+	mock := quota.NewMockManager(ctrl)
+	mock.EXPECT().GetUser(ctx, nil).Return(want)
+
+	qm := NewCachedManager(mock, batchSize, maxEntries)
+	if got := qm.GetUser(ctx, nil /* req */); got != want {
+		t.Errorf("GetUser() = %v, want = %v", got, want)
+	}
+}
+
+func TestCachedManager_PeekTokens(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	tests := []struct {
+		desc       string
+		wantTokens map[quota.Spec]int
+		wantErr    error
+	}{
+		{
+			desc: "success",
+			wantTokens: map[quota.Spec]int{
+				{Group: quota.Global, Kind: quota.Read}: 10,
+				{Group: quota.Global, Kind: quota.Read}: 11,
+			},
+		},
+		{
+			desc:    "error",
+			wantErr: errors.New("llama ate all tokens"),
+		},
+	}
+
+	ctx := context.Background()
+	for _, test := range tests {
+		mock := quota.NewMockManager(ctrl)
+		mock.EXPECT().PeekTokens(ctx, specs).Return(test.wantTokens, test.wantErr)
+
+		qm := NewCachedManager(mock, batchSize, maxEntries)
+		tokens, err := qm.PeekTokens(ctx, specs)
+		if diff := pretty.Compare(tokens, test.wantTokens); diff != "" {
+			t.Errorf("%v: post-PeekTokens() diff (-got +want):\n%v", test.desc, diff)
+		}
+		if err != test.wantErr {
+			t.Errorf("%v: PeekTokens() returned err = %v, want = %v", test.desc, err, test.wantErr)
+		}
+	}
+}
+
+// TestCachedManager_DelegatedMethods tests all delegated methods that have a single error return.
+func TestCachedManager_DelegatedMethods(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	ctx := context.Background()
+	tokens := 5
+	for _, want := range []error{nil, errors.New("llama ate all tokens")} {
+		mock := quota.NewMockManager(ctrl)
+		qm := NewCachedManager(mock, batchSize, maxEntries)
+
+		mock.EXPECT().PutTokens(ctx, tokens, specs).Return(want)
+		if err := qm.PutTokens(ctx, tokens, specs); err != want {
+			t.Errorf("PutTokens() returned err = %v, want = %v", err, want)
+		}
+
+		mock.EXPECT().ResetQuota(ctx, specs).Return(want)
+		if err := qm.ResetQuota(ctx, specs); err != want {
+			t.Errorf("ResetQuota() returned err = %v, want = %v", err, want)
+		}
+	}
+}
+
+func TestCachedManager_GetTokens_CachesTokens(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	ctx := context.Background()
+	tokens := 3
+	mock := quota.NewMockManager(ctrl)
+	mock.EXPECT().GetTokens(ctx, tokens+batchSize, specs).Times(2).Return(nil)
```

Comment:
To make the test less brittle, this should just expect that GetTokens is called with at least `tokens` or `batchSize`, whichever is greater (that's all that the documentation guarantees).